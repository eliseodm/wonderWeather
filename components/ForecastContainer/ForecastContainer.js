import { StyleSheet, Text, ScrollView, View } from 'react-native';
import React from 'react';
import ForecastDay from '../ForecastDay/ForecastDay';
import { dayNames } from "../../constants";


function ForecastContainer({ forecastInfo }) {
    return(
        <View style={styles.forecastContainer}>
            <Text style={styles.subtitle}>Proximos 5 dias</Text>
            <ScrollView showsVerticalScrollIndicator={false}>
                <View>
                    {
                        forecastInfo.map((day) => {
                            return(
                                <ForecastDay
                                    key={day.dayNum}
                                    minTemp={day.minTemp}
                                    maxTemp={day.maxTemp}
                                    icon={day.icon}
                                    day={dayNames[day.dayNum]}
                                />
                            )
                        })
                    }
                </View>
            </ScrollView>
        </View>
    )
}

const styles = StyleSheet.create({
    subtitle:{
        fontSize: 24,
        marginVertical: 12,
        marginLeft: 4,
        color: '#e96e50',
    },
    forecastContainer: {
        marginTop: 16,
        height: 300,
    }
})

export default ForecastContainer;