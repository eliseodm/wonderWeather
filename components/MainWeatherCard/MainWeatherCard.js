import { StyleSheet, View, Text, Image } from 'react-native';
import React from 'react';


function MainWeatherCard({current}) {
    return (
        <>
            <Text style={styles.city}>{current.city}</Text>
            <Text style={styles.country}>{current.country}</Text>
            <View style={styles.current}>
                <Image
                    style={styles.largeIcon}
                    source={{
                    uri: `https://openweathermap.org/img/wn/${current.icon}@4x.png`,
                    }}
                />
                <Text style={styles.currentTemp}>{Math.round(current.temp)}°C</Text>
                <Text style={styles.maxTemp}>Max {Math.round(current.maxTemp)}°C</Text>
                <Text style={styles.minTemp}>Min {Math.round(current.minTemp)}°C</Text>
            </View>
        </>
    );
}

export default MainWeatherCard;

const styles = StyleSheet.create({
    city: {
        width: '100%',
        textAlign: 'center',
        fontSize: 32,
        color: '#e96e50'
    },
    country: {
        width: '100%',
        textAlign: 'center',
        fontSize: 24,
        color: '#d6a99e'
    },
    current: {
        flexDirection: 'column',
        alignItems: 'flex-end',
        alignContent: 'center',
    },
    largeIcon: {
        width: 250,
        height: 200,
      },
      currentTemp: {
        fontSize: 32,
        fontWeight: 'bold',
        textAlign: 'center',
      },
      maxTemp: {
        fontSize: 16,
        fontWeight: 'bold',
        textAlign: 'center',
      },
      minTemp: {
        fontSize: 16,
        fontWeight: 'bold',
        textAlign: 'center',
      },  


})