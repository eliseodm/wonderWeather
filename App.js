import React, { useState, useEffect, useCallback } from 'react';
import {
  ActivityIndicator,
  SafeAreaView,
  StyleSheet,
  Text,
} from 'react-native';

import MainWeatherCard from './components/MainWeatherCard/MainWeatherCard';
import ForecastContainer from './components/ForecastContainer/ForecastContainer';
import SelectedCityModal from './components/SelectedCityModal/SelectedCityModal';

import usePublicIp from "./hooks/usePublicIp";
import useLocationInfo from "./hooks/useLocationInfo";
import useForecastInfo from './hooks/useForecastInfo';
import { listOfCities } from "./constants";

const App = () => {
  const [citiesList, setCitiesList] = useState(listOfCities);
  const [selectedCity, setSelectedCity] = useState({});
  const { publicIpV4 } = usePublicIp();
  const { locationData } = useLocationInfo(publicIpV4);
  const { forecastInfo, todayInfo, loading } = useForecastInfo(
    selectedCity.lat,
    selectedCity.lon
  );
  
  const isCurrentLocation = locationData && locationData.city === selectedCity.city;

  useEffect(() => {
    if (
      locationData !== null &&
      locationData.city &&
      citiesList[0].city !== locationData.city
    ) {
      setCitiesList([locationData, ...citiesList]);
      setSelectedCity(locationData);
    }
  }, [locationData, setCitiesList, citiesList]);

  const onSelectCity = useCallback((option) => {
    setSelectedCity(option);
  }, []);

  if(loading){
    return <SafeAreaView style={styles.loading}>
              <ActivityIndicator testID="App.ActivityIndicator" size="large" />
          </SafeAreaView>;
  }

  return (
    <SafeAreaView style={styles.container}>
      <Text style={styles.appTitle}>Wonder Weather</Text>
      <MainWeatherCard
        isCurrentLocation={isCurrentLocation}
        cityInfo={{ ...todayInfo, ...selectedCity }}
      />
      <ForecastContainer
        forecastInfo={forecastInfo}
      />
      <SelectedCityModal
        onSelectCity={onSelectCity}
        citiesList={citiesList}
      />
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    alignItems: 'center',
    justifyContent: 'flex-start',
  },
  appTitle: {
    fontSize: 16,
    fontWeight: 'bold',
    marginTop: 10,


  },
  loading: {
    flex: 1,
    backgroundColor: '#fff',
    alignItems: 'center',
    justifyContent: 'center',
  },
});

export default App;
