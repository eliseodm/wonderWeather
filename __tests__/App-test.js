import 'react-native';
import React from 'react';
import { render, screen } from '@testing-library/react-native';
import { toHaveTextContent } from '@testing-library/jest-native/extend-expect';
import { mock_cityInfo, mock_forecastInfo } from "../__mockData__/index"
import App from '../App';
import MainWeatherCard from '../components/MainWeatherCard/MainWeatherCard';
import ForecastContainer from '../components/ForecastContainer/ForecastContainer';

it('Generates a Snapshot of <App/>', () => {
  const { toJSON } =render(<App />);
  expect(toJSON()).toMatchSnapshot()
});

const mainComponent = <App />;

//show ActivityIndicator before fetch data
it('Show ActivityIndicator before fetch fata', () => {
  const { getByTestId } = render(mainComponent);
  
  expect(getByTestId('App.ActivityIndicator')).toBeTruthy();
});

//show current location info
describe('Show current location info', () => {
  it('City', () => {
    const { getByTestId } = render(
      <MainWeatherCard  cityInfo={mock_cityInfo} isCurrentLocation={true} />
    )
    expect(getByTestId("MainWeatherCard.city")).toHaveTextContent("San Miguel de TucumÃ¡n");
  })
  
  it('Country', () => {
    const { getByTestId } = render(
      <MainWeatherCard  cityInfo={mock_cityInfo} isCurrentLocation={true} />
    )
    expect(getByTestId("MainWeatherCard.country")).toHaveTextContent("Argentina");
  })
})
  it('Temp', () => {
    const { getByTestId } = render(
      <MainWeatherCard  cityInfo={mock_cityInfo} isCurrentLocation={true} />
    )
    expect(getByTestId("MainWeatherCard.temp")).toHaveTextContent("24");
  })

//show current location next 5 days forecast
  it('Get current location next 5 days forecast', () => {
    render(<ForecastContainer forecastInfo={mock_forecastInfo} />)
    expect(screen.getAllByTestId("ForecastContainer.days").length).toBe(5);
  })
