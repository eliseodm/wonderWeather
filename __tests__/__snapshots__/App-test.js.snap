// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generates a Snapshot of <App/> 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <ActivityIndicator
    size="large"
    testID="App.ActivityIndicator"
  />
</RCTSafeAreaView>
`;